frontend/kubecloud/src/components/deploy/Step2AssignNodes.vue

```diff
@@ -149,6 +133,79 @@ function nodeLabel(node: any) {
   if (!node) return '';
   return `Node ${node.nodeId}`;
 }
+
+// Calculate allocated resources for all VMs except the current one being assigned
+function getAllocatedResources(excludeVmIndex?: number) {
+  const allocations: Record<number, { ram: number; storage: number }> = {};
+  
+  props.allVMs.forEach((vm, index) => {
+    // Skip the current VM being assigned to avoid self-blocking
+    if (excludeVmIndex !== undefined && index === excludeVmIndex) return;
+    
+    if (vm.node != null) {
+      if (!allocations[vm.node]) {
+        allocations[vm.node] = { ram: 0, storage: 0 };
+      }
+      allocations[vm.node].ram += vm.ram;
+      allocations[vm.node].storage += vm.disk || 0;
+    }
+  });
+  
+  return allocations;
+}
+
+function getNodeAvailableResources(node: NormalizedNode, excludeVmIndex?: number) {
+  const allocations = getAllocatedResources(excludeVmIndex);
+  const allocated = allocations[node.nodeId] || { ram: 0, storage: 0 };
+  return {
+    cpu: node.cpu || 0, // CPU is always available (can be oversubscribed)
+    ram: Math.max(0, (node.available_ram || 0) - allocated.ram),
+    storage: Math.max(0, (node.available_storage || 0) - allocated.storage)
+  };
+}
+
+function getAvailableNodesForVM(vmIndex: number) {
+  const vm = props.allVMs[vmIndex];
+  if (!vm) return [];
+  
+  return props.availableNodes.filter(node => {
+    const available = getNodeAvailableResources(node, vmIndex);
+    // CPU check: ensure node has minimum required CPU (but allow oversubscription)
+    const hasSufficientCpu = (node.cpu || 0) >= vm.vcpu;
+    
+    return hasSufficientCpu && 
+           available.ram >= vm.ram && 
+           available.storage >= (vm.disk || 0);
+  });
+}
+
+// Validate existing VM assignments on mount and clear invalid ones
+onMounted(() => {
+  props.allVMs.forEach((vm, vmIndex) => {
+    if (vm.node != null) {
+      // Find the assigned node
+      const assignedNode = props.availableNodes.find(node => node.nodeId === vm.node);
+      
+      if (!assignedNode) {
+        // Node doesn't exist anymore, clear assignment
+        props.onAssignNode(vmIndex, null);
+        return;
+      }
+      
```

Comment:
This operation is costly. On every node selection, you calculate the resources for all the nodes. meaning that if you have 5 nodes, you'll calculate the resources at least 16 times. 

Instead of calculating resources by node, why don't we calculate by VM? This way, we'll enhance the performance to be O(n) instead of O(n)^2.

Reply:
Code snippet above is wrong. check below for the correct one.

https://github.com/codescalers/kubecloud/blob/cab3f60847b9d4a6d1f3047b371479a4609c9fba/frontend/kubecloud/src/components/deploy/Step2AssignNodes.vue#L167-L181